Add documentation for Heft (#16)
microsoft/rushstack.io-website: pages/heft/core_plugins.md

---
layout: page
title: Core plugins
navigation_source: docs_nav
---

| Plugin  | Description |
| ------------- | ------------- |
| [ApiExtractorPlugin](https://github.com/microsoft/rushstack/blob/master/apps/heft/src/plugins/ApiExtractorPlugin/ApiExtractorPlugin.ts) | Implements the [api-extractor]({% link pages/heft_tasks/api-extractor.md %}) task |

> | [BasicConfigureWebpackPlugin](https://github.com/microsoft/rushstack/blob/master/apps/heft/src/plugins/Webpack/BasicConfigureWebpackPlugin.ts) | Used to configure the [webpack]({% link pages/heft_tasks/webpack.md %}) task by loading a static **webpack.config.js** file (instead of building a dynamic configuration at runtime) |

Reviewer: ```suggestion
| [BasicConfigureWebpackPlugin](https://github.com/microsoft/rushstack/blob/master/apps/heft/src/plugins/Webpack/BasicConfigureWebpackPlugin.ts) | Used to configure the [webpack]({% link pages/heft_tasks/webpack.md %}) task by loading a static **webpack.config.js** file specified in a configuration file, instead of building a dynamic configuration at runtime |
```

Author: From https://github.com/microsoft/rushstack/pull/2037#discussion_r457086267 we decided to standardize the **webpack.config.js** filename.

> | [CleanPlugin](https://github.com/microsoft/rushstack/blob/master/apps/heft/src/plugins/CleanPlugin.ts) | Deletes temporary files from a previous build |

Reviewer: ```suggestion
| [CleanPlugin](https://github.com/microsoft/rushstack/blob/master/apps/heft/src/plugins/CleanPlugin.ts) | Deletes files from a previous build |
```

| [CopyStaticAssetsPlugin](https://github.com/microsoft/rushstack/blob/master/apps/heft/src/plugins/CopyStaticAssetsPlugin.ts) | Implements the [copy-static-assets]({% link pages/heft_tasks/copy-static-assets.md %}) task |
| [JestPlugin](https://github.com/microsoft/rushstack/blob/master/apps/heft/src/plugins/JestPlugin/JestPlugin.ts) | Implements the [jest]({% link pages/heft_tasks/jest.md %}) task |
| [PackageJsonConfigurationPlugin](https://github.com/microsoft/rushstack/blob/master/apps/heft/src/plugins/PackageJsonConfigurationPlugin.ts) | Disables the [typescript]({% link pages/heft_tasks/typescript.md %}) tasks's symlinking optimization for published projects, because publishing workflows do not handle symlinks properly |

> | [ProjectActionConfigurationFilesPlugin](https://github.com/microsoft/rushstack/blob/master/apps/heft/src/plugins/ActionConfigurationLoaders/ProjectActionConfigurationFilesPlugin.ts)  | Reads project-specific config files for Heft "actions" such as **clean.json**, **typescript.json**, **copy-static-assets.json**, and **webpack.json** |

Reviewer: ```suggestion
| [ProjectActionConfigurationFilesPlugin](https://github.com/microsoft/rushstack/blob/master/apps/heft/src/plugins/ActionConfigurationLoaders/ProjectActionConfigurationFilesPlugin.ts)  | Reads project-specific config files such as **clean.json**, **typescript.json**, **copy-static-assets.json**, and **webpack.json** |
```

The use of "actions" in these plugins is a relic from a previous design.

Author: Fixed

> | [RushActionConfigurationFilesPlugin](https://github.com/microsoft/rushstack/blob/master/apps/heft/src/plugins/ActionConfigurationLoaders/RushActionConfigurationFilesPlugin.ts)  | Reads Rush monorepo config files for Heft "actions" such as **clean.json**, **typescript.json**, **copy-static-assets.json**, and **webpack.json** |

Reviewer: ```suggestion
| [RushActionConfigurationFilesPlugin](https://github.com/microsoft/rushstack/blob/master/apps/heft/src/plugins/ActionConfigurationLoaders/RushActionConfigurationFilesPlugin.ts)  | Reads Rush monorepo-wide config files such as **clean.json**, **typescript.json**, **copy-static-assets.json**, and **webpack.json** |
```

Author: Fixed
